### PR TITLE
[ffmpeg/mfx-dispatch] Add mfx-dispatch port, add qsv support and amf support for ffmpeg 4.4.3

### LINKED ISSUE #9272: [ffmpeg] add amd/amf, qsv hard encoding and openh264 encoding support

### PR DESCRIPTION
**Pull request by reitowo**
**Describe the pull request**

## Explain

  - Because the qsv feature of ffmpeg depends on mfx-dispatch, I cannot separate this 2 ports into 2 PRs.
  - The introduced ffmpeg's patch 0023 is there because we still on version 4.4.3 (and no one is going to upgrade port to 5.x?) and I also didn't check whether ffmpeg upstream fixed this initialization problem of qsv: 
https://trac.ffmpeg.org/ticket/7933#comment:3
https://github.com/InitialForce/FFmpeg/commit/338fbcd5bba1de0e1b3e3bad8985eee2fdfbeca1
       - After the patch is applied, qsv will work properly under multi-monitor situation or something else prevented it from working.
       - Remove this patch **only** when upstream fix this problem, otherwise the qsv feature will be broken

## PR

- #### What does your PR fix?
  Fixes #9272  #20415 #9247

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  windows, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

### CODE REVIEW

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -592,6 +592,10 @@ mesa:x64-osx=fail
 mfl:x64-linux=skip
 mfl:x64-osx=skip
 mfl:arm64-osx=skip
+
+mfx-dispatch:x64-windows=pass
+mfx-dispatch:x86-windows=pass
+
```
Comment (Adela0814) [inline scripts/ci.baseline.txt:598]: Unnecessary, please remove this.

File: ports/mfx-dispatch/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}" 
+)
+
+vcpkg_cmake_install()
+vcpkg_copy_pdbs()
+vcpkg_fixup_pkgconfig()
+  
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+configure_file("${SOURCE_PATH}/LICENSE" "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" COPYONLY)
```
Comment (Adela0814) [inline ports/mfx-dispatch/portfile.cmake:24]: Please add a new line at the end of the file.

File: ports/mfx-dispatch/fix-unresolved-symbol.patch [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+===================================================================
+diff --git a/src/mfx_driver_store_loader.cpp b/src/mfx_driver_store_loader.cpp
+--- a/src/mfx_driver_store_loader.cpp	(revision 388559e9e8234eb0989e1598a9beea4035a04132)
++++ b/src/mfx_driver_store_loader.cpp	(revision 5ebef171699530ca01594a5cef10a68811f4d105)
+@@ -24,6 +24,8 @@
+ #include "mfx_dispatcher_log.h"
+ #include "mfx_load_dll.h"
+ 
++#pragma comment(lib, "Ole32.lib")
++
```
Comment (Adela0814) [inline ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: Could you please create a PR to upstream.

File: ports/ffmpeg/vcpkg.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "ffmpeg",
   "version": "4.4.3",
+  "port-version": 2,
```
Comment (Adela0814) [inline ports/ffmpeg/vcpkg.json:4]:
```suggestion
  "port-version": 1,
```

**Review by Adela0814 — CHANGES_REQUESTED** [review]

File: ports/mfx-dispatch/fix-unresolved-symbol.patch (hunk 3 above)
Comment (cnSchwarzer) [reply ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: OK, I'll try to request a PR at the upstream. Can we approve this PR first, and I will start another PR when the mfx-dispatch upstream accepted this fix?

Comment (cnSchwarzer) [reply ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: https://github.com/lu-zero/mfx_dispatch/pull/92#issue-1495645126

Comment (Adela0814) [reply ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: Thanks, Let's wait for this PR to be approved.

Comment (cnSchwarzer) [reply ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: @Adela0814 I find out the upstream seems inactive for months, and I guess the PR won't get merged in short time, could we merge this port and I promise to keep in touch with upstream, start another port update PR if upstream accepted my PR？

Comment (cnSchwarzer) [reply ports/mfx-dispatch/fix-unresolved-symbol.patch:72]: Because some repo of mine depends on this feature, and I can't let everyone use vcpkg registry.

**Review by Adela0814 — DISMISSED** [review]

**Comment by Adela0814** [comment]
We checked that the upstream has been inactive for a long time, tagged `'reviewed'` in order not to block this PR, what do you think? @dan-shaw

**Comment by dan-shaw** [comment]
I don't think we can accept the `ffmpeg` patch, as it makes significant changes to source. It would be better to wait for a version update instead. I think the `mfx-dispatch` changes are okay, as upstream has no update for over a year.

**Comment by cnSchwarzer** [comment]
The patch 0023 is just like all other ffmpeg patches, it just fix an issue with qsv feature.

**Comment by cnSchwarzer** [comment]
Anyway I started #28387 to upgrade ffmpeg to 5.1.2, but it will hang on linking and I have no idea what went wrong.

**Comment by cnSchwarzer** [comment]
As the `amd-amf` port was merged into master, I also added support `amf` for ffmpeg. It is working.
image

**Comment by cnSchwarzer** [comment]
The vcpkg-ci-ffmpeg need to remove feature `avresample` because this library is deprecated in ffmpeg 5.x, I've made change to the vcpkg.json but the ci seems not using this changed version.

**Comment by cnSchwarzer** [comment]
Another fact is ffmpeg 5.x is api breaking, it removed avresample feature, so I think existing user and ports which depends on 4.4.3 may have problem upgrading to 5.x, and this patch 0023 is essential for 4.x user to continue use this version.. @dan-shaw

**Review by dan-shaw — APPROVED** [review]

File: ports/mfx-dispatch/fix-unresolved-symbol.patch [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+--- a/CMakeLists.txt	(revision 7e4d221c36c630c1250b23a5dfa15657bc04c10c)
++++ b/CMakeLists.txt	(revision 5ebef171699530ca01594a5cef10a68811f4d105)
+@@ -18,8 +18,7 @@
+   src/main.cpp
+   src/mfx_dispatcher.cpp
+   src/mfx_dispatcher_log.cpp
+-  src/mfx_function_table.cpp
+-)
++  src/mfx_function_table.cpp)
```
Comment (dan-shaw) [inline ports/mfx-dispatch/fix-unresolved-symbol.patch:18]: Can we remove this, to minimize the patch?

File: ports/mfx-dispatch/fix-unresolved-symbol.patch [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+===================================================================
+diff --git a/libmfx.pc.cmake b/libmfx.pc.cmake
+--- a/libmfx.pc.cmake	(revision 7e4d221c36c630c1250b23a5dfa15657bc04c10c)
++++ b/libmfx.pc.cmake	(revision 388559e9e8234eb0989e1598a9beea4035a04132)
+@@ -9,6 +9,6 @@
+ Requires:
+ Requires.private:
+ Conflicts:
+-Libs: -L${libdir} -lsupc++ ${libdir}/libmfx.a
++Libs: -L${libdir} -lsupc++ ${libdir}/libmfx.lib
```
Comment (dan-shaw) [inline ports/mfx-dispatch/fix-unresolved-symbol.patch:56]: Can we add something like `-lmfx` to make it platform independent?

**Review by dan-shaw — DISMISSED** [review]

**Comment by cnSchwarzer** [comment]
@dan-shaw Could you please merge this today? It helps a lot🙏🏻 And wait for current ci is meaningless because nothing significant changed

**Pull request merged**